### ISSUE TITLE
TaskManager Launchers do not match correctly with Windows [kmail, gentoo example]

### ISSUE THREAD

**Issue by alexjp** [psifidotos/nowdock-plasmoid#24]
Hi,

I am having a weird issue with kmail launcher. I made a video showing what is happening: https://youtu.be/rBOmJDK5m08

Kmail launch icon seems to be always in every activity, even though its not listed as "in all activities".
Also notice that when I "Show a launcher when not running" to the kmail app, It is configured as "not show in all activities" already, like if it already had information about that app.

I looked as plasma configuration files and nowdock launcher listing. Your format is easy to understand -> activity;number of launchers;launchers.

Can you try to reproduce ? I was able to fix it, by creating a custom kmail-2-launcher.desktop shortcut.

Thanks !

---

https://bugs.gentoo.org/show_bug.cgi?id=593084
https://bugs.kde.org/show_bug.cgi?id=368425

**Comment by psifidotos** [comment]
> It is configured as "not show in all activities" already, like if it already had information about that app.

There are two states for each launcher either "In all activities" or "according to activities". for example I use 4 activities and the dolphin launcher is used in two of them. The fact that when you add the kmail launcher and indicates that is not on all activities, it means that is finding it in another activity so the code understands that this must not be added as "OnAllActivities" because this would break the user workflow.  The problem is when you hit "On All Activities" it should make this globally available but this doesnot happen... I will investigate it...

Notice: If you want to improve a bit your visuals(big icons from Default panel) you could use the Now Dock Panel and plasmoid (master branches). The Now Dock Panel has now its own configuration settings window.. You could set it to not show a background of its own and use as a zoom factor 1, that would stop the animation hovering and set anchoring at top... Small problem I dont support yet spacers, in order to have other items at top and other at bottom...

**Comment by psifidotos** [comment]
Ok ... my mistake ... I observed your video again... The behavior is ok... this is not a bug...

The question is ... Where do you want your kmail launcher? :) in what activities?

**Comment by alexjp** [comment]
Thanks for the reply:

> The question is ... Where do you want your kmail launcher? :) in what activities?

I want the launcher in the PIM Activity.

> The fact that when you add the kmail launcher and indicates that is not on all activities, it means that is finding it in another activity so the code understands that this must not be added as "OnAllActivities" because this would break the user workflow. 

It isnt, the kmail launcher is not anywhere ( i checked the plasma rc file and the launchers line of nowdock )  ... I think its relative to that my system kmail desktop file executes the command 'kmail -qwindowtitle %c %u'. My custom kmail desktop file just executes 'kmail'. I triedc to put my custom kmail desktop file to execute the same as the system file, and the problem reappeared.

**Comment by alexjp** [comment]
I am beggining to wonder if this is an upstream bug: 

for example, on gentoo and openSUSE, none of the chrome apps launchers work properly.
on arch, they work properly.

and this is using the same versions of plasma / frameworks / qt

something the way the taskmanager lib detects the launcher and the running window.

**Comment by psifidotos** [comment]
from the Now Dock rc, can you show me your configuration line for launchers? Use both cases, the one working and the one that it isnt...

**Comment by psifidotos** [comment]
And you are sure that the kmail launcher is not also in any stopped activity?

**Comment by alexjp** [comment]
non-working (kmail is actually working because i use my custom launcher with edited command line, but chrome apps are not working): 

[Containments][290][Applets][292]
immutability=1
plugin=org.kdelook.nowdock

[Containments][290][Applets][292][Configuration][ConfigDialog]
DialogHeight=765
DialogWidth=878

[Containments][290][Applets][292][Configuration][General]
dotsOnActive=true
launchers=multi;def93e1d-9350-4ae9-b0d1-0724bca097c3;1;file:///usr/share/applications/chromium-browser-chromium.desktop;*;2;file:///usr/share/applications/org.kde.dolphin.desktop;file:///usr/share/applications/org.kde.konsole.desktop;8be9f0c8-04de-4a66-b157-2b5ef83caf5b;3;file:///usr/share/applications/vlc.desktop;file:///usr/share/applications/smplayer.desktop;file:///home/alex/.local/share/applications/chrome-blpcfgokakmgnkcojhhkbfbldkacnbeo-Default.desktop;db184444-49a4-4417-af83-919bfc8d32a0;9;file:///home/alex/.local/share/applications/chrome-knipolnnllmklapflnccelgolnpehhpl-Default.desktop;file:///home/alex/.local/share/applications/chrome-oonccmmafcaodljbcgobdbknmbljiafh-Default.desktop;file:///home/alex/.local/share/applications/chrome-dlppkpafhbajpcmmoheippocdidnckmm-Default.desktop;file:///home/alex/.local/share/applications/chrome-blpcfgokakmgnkcojhhkbfbldkacnbeo-Default.desktop;file:///home/alex/.local/share/applications/chrome-lgnggepjiihbfdbedefdhcffnmhcahbm-Default.desktop;file:///home/alex/.local/share/applications/chrome-celnaknmndcdcjcagffhbhciignkeokb-Default.desktop;file:///home/alex/.local/share/applications/chrome-hbdpomandigafcibbmofojjchbcdagbl-Default.desktop;file:///home/alex/.local/share/applications/chrome-mhofgldekhpmogcppcmggpfiejhmlpne-Default.desktop;file:///home/alex/.local/share/applications/chrome-clhhggbfdinjmjhajaheehoeibfljjno-Default.desktop;7b47ef4c-488c-4cb5-bc6a-5fb112885ab3;4;file:///home/alex/.local/share/applications/chrome-akjnfnchjccpnhmdliflclkdjplampka-Default.desktop;file:///home/alex/.local/share/applications/chrome-amneadmcchmfcpmojkidfcjhmmgakapa-Default.desktop;file:///home/alex/.local/share/applications/chrome-enmjggaoopkfkjncogpnhnpaoifpapki-Default.desktop;file:///home/alex/.local/share/applications/chrome-bechbfmlpenelokakfbonfmmgecpikhg-Default.desktop;69fc8096-036d-47d8-ab04-e55542b0095e;6;file:///usr/share/applications/org.kde.korganizer.desktop;file:///home/alex/.local/share/applications/chrome-pkclgpgponpjmpfokoepglboejdobkpl-Default.desktop;file:///home/alex/.local/share/applications/chrome-ppbiamkfhlcpgijakdgbbffikmkkofnl-Default.desktop;file:///home/alex/.local/share/applications/chrome-ejjicmeblgpmajnghnpcppodonldlgfn-Default.desktop;file:///home/alex/.local/share/applications/chrome-hmjkmjkepdijhoojdojkdfohbdgmmhki-Default.desktop;file:///home/alex/.local/share/applications/org.kde.kmail-2.desktop
plasmoidPosition=Top
showBarLine=false
threeColorsWindows=true
transparentPanel=true
zoomLevel=0

working:

[Containments][290][Applets][292]
immutability=1
plugin=org.kdelook.nowdock

[Containments][290][Applets][292][Configuration][ConfigDialog]
DialogHeight=765
DialogWidth=878

[Containments][290][Applets][292][Configuration][General]
dotsOnActive=true
launchers=multi;*;3;file:///usr/share/applications/org.kde.dolphin.desktop;file:///usr/share/applications/org.kde.konsole.desktop;file:///home/alex/.local/share/applications/org.kde.kmail-2.desktop;8be9f0c8-04de-4a66-b157-2b5ef83caf5b;3;file:///usr/share/applications/vlc.desktop;file:///usr/share/applications/smplayer.desktop;file:///home/alex/.local/share/applications/chrome-blpcfgokakmgnkcojhhkbfbldkacnbeo-Default.desktop;db184444-49a4-4417-af83-919bfc8d32a0;9;file:///home/alex/.local/share/applications/chrome-celnaknmndcdcjcagffhbhciignkeokb-Default.desktop;file:///home/alex/.local/share/applications/chrome-knipolnnllmklapflnccelgolnpehhpl-Default.desktop;file:///home/alex/.local/share/applications/chrome-dlppkpafhbajpcmmoheippocdidnckmm-Default.desktop;file:///home/alex/.local/share/applications/chrome-lgnggepjiihbfdbedefdhcffnmhcahbm-Default.desktop;file:///home/alex/.local/share/applications/chrome-oonccmmafcaodljbcgobdbknmbljiafh-Default.desktop;file:///home/alex/.local/share/applications/chrome-clhhggbfdinjmjhajaheehoeibfljjno-Default.desktop;file:///home/alex/.local/share/applications/chrome-hbdpomandigafcibbmofojjchbcdagbl-Default.desktop;file:///home/alex/.local/share/applications/chrome-mhofgldekhpmogcppcmggpfiejhmlpne-Default.desktop;file:///home/alex/.local/share/applications/chrome-blpcfgokakmgnkcojhhkbfbldkacnbeo-Default.desktop;7b47ef4c-488c-4cb5-bc6a-5fb112885ab3;4;file:///home/alex/.local/share/applications/chrome-akjnfnchjccpnhmdliflclkdjplampka-Default.desktop;file:///home/alex/.local/share/applications/chrome-amneadmcchmfcpmojkidfcjhmmgakapa-Default.desktop;file:///home/alex/.local/share/applications/chrome-enmjggaoopkfkjncogpnhnpaoifpapki-Default.desktop;file:///home/alex/.local/share/applications/chrome-bechbfmlpenelokakfbonfmmgecpikhg-Default.desktop;69fc8096-036d-47d8-ab04-e55542b0095e;5;file:///home/alex/.local/share/applications/chrome-pkclgpgponpjmpfokoepglboejdobkpl-Default.desktop;file:///home/alex/.local/share/applications/chrome-ppbiamkfhlcpgijakdgbbffikmkkofnl-Default.desktop;file:///home/alex/.local/share/applications/chrome-ejjicmeblgpmajnghnpcppodonldlgfn-Default.desktop;file:///home/alex/.local/share/applications/chrome-hmjkmjkepdijhoojdojkdfohbdgmmhki-Default.desktop;file:///usr/share/applications/korganizer-import.desktop
plasmoidPosition=Top
showBarLine=false
threeColorsWindows=true
transparentPanel=true
zoomLevel=0

( they are almost the same, because i tried copying the working plasma\* files and kactivity\* files from the working system to the non working systems )

**Comment by alexjp** [comment]
This is an upstream bug, sorry: I just tested with normal taskmanager plasmoid. NowDock is just doing exactly what normal kde taskmanager does, with kmail and chrome apps.

Feel free to close it.

**Comment by psifidotos** [comment]
Ok ...

**Comment by Steinhagen** [comment]
Is there a bug raised upstream for this problem?

**Comment by alexjp** [comment]
> Is there a bug raised upstream for this problem?

Don't think so, at least in my quick search

**Comment by alexjp** [comment]
To anyone that has this problem, i was able to fix it by changing the system chromium way of launching and wm_class.

gentoo and opensuse make chromium have class: "chromium-browser-chromium"
archlinux has wmclass: "chromium".

on the gentoo system, i was able to make a chrome app have the xprop with wmclass:
WM_CLASS(STRING) = "crx_ppbiamkfhlcpgijakdgbbffikmkkofnl", "chromium"

and it works now.

Its weird why it works with wmclass "chromium" and not "chromium-browser-chromium", since on gentoo, its consistent in all windows having "chromium-browser-chromium" wmclass.

**Comment by alexjp** [comment]
Note: submitted bug upstream to gentoo: https://bugs.gentoo.org/show_bug.cgi?id=593084

**Comment by alexjp** [comment]
Also submitted to https://bugs.kde.org/show_bug.cgi?id=368425

**Comment by psifidotos** [comment]
Very good job!!!

Alex can we find a good title for this? "Launchers not matching correctly with Windows" ,maybe?

**Comment by alexjp** [comment]
Yes, maybe a better title. This is kinda hard to describe. Expecially to the gentoo mantainers, who might not even use KDE.

BTW, dunno where better to tell you this, but also reported the activity sidebar problem on kde forums: https://forum.kde.org/viewtopic.php?f=285&t=135869

**Comment by psifidotos** [comment]
I have also added the links to your first message in order to be easily accessible!
I hope you are ok with this!

**Comment by alexjp** [comment]
yep, no problem :)

**Comment by psifidotos** [comment]
> BTW, dunno where better to tell you this, but also reported the activity sidebar problem on kde forums..

This is also a very good job!!! :-) happy for you, Alex!!
the activity sidebar can be also opened as kde bug under plasmashell->activity switcher, maybe a feature request which would also have a link for the forum link.
The maintainer, I think is Ivan and something tells me there will be a very short response... :)

**Comment by psifidotos** [comment]
Concerning the issue of this topic, I think I have the same problem under openSuse with LibreOffice Writer....

**Comment by alexjp** [comment]
It works were on Gentoo:

Can you try removing the %U at the end of the .desktop command line ?

**Comment by psifidotos** [comment]
@alexjp  for my libreoffice writer which has the issue...
my desktop file is,
**/usr/share/applications/writer.desktop**
inside there are the lines:
**StartupWMClass=libreoffice-writer**
**Exec=libreoffice --writer**

any ideas?

**Comment by psifidotos** [comment]
it is not strange that libtaskmanager can not identify the libreoffice-writer window with its launcher... :)

**Comment by alexjp** [comment]
try to rename your desktop file to libreoffice-writer

also check with xprop the name of the WM_CLASS. I think the desktop file has to match with it.

**Comment by psifidotos** [comment]
:( ... didnt help in my case....

**Comment by alexjp** [comment]
this is my gentoo desktop file that is working: http://pastebin.com/pWpmsMtm

also, my xprop:  http://pastebin.com/XQPcfkqT

can you show me both those files ?

Also did you removed the launchers from nowdock and inserted them again ? sometimes nowdock creates weird launchers ( not desktop based, and even include the icon in the plasma rc file )

**Comment by kupiqu** [comment]
I might be the only one using konqueror yet but for the case it is useful for someone out there:

Konqueror suffers this same issue, which I workarounded by copying konqbrowser.desktop to konqueror.desktop and replacing in it
`Exec=kfmclient openProfile webbrowsing
`
by
`Exec=konqueror
`

**Comment by psifidotos** [comment]
my renamed desktop file: http://pastebin.com/4Csvh7s2
my xprop: http://pastebin.com/thBS0jvz

yes, I removed the launchers( when I renamed the desktop file these launcher became a question mark) and readded for libreoffice writer

**Comment by alexjp** [comment]
weird ... seems all ok, seems to point to libreoffice-writer.desktop.

can you change the : Exec=libreoffice --writer %U to Exec=libreoffice --writer %F ?

**Comment by psifidotos** [comment]
@alexjp, this fixed it!!!! Thank u very much!!!!

Did you have any response to where we report such cases? plasma or distribution?

**Comment by alexjp** [comment]
> Did you have any response to where we report such cases? plasma or distribution?

I reported it on kde, and Eike Hein said it should be the app or distribution responsability. Quoting him:

> It's the distro's or app dev's duty to make the window meta data and menu id (.desktop name) align so task managers can map one to the other. The KMail devs were informed about their bug some time ago ... there's a ticket for KMail somewhere.

In case of libreoffice, i guess it should be the distribution, since its different from mine. Till now, only case i catched that is cross distribution is the kmail case. Even chromium is working out of the box on arch linux.

I guess it was never visible because nobody made use of taskmanager tracking launchers. Maybe put some info on front page README, something like: " If the dock isnt tracking correctly a launcher or chromium app, complaint to your distribution to fix the desktop files, see workarounds for common fixes" and make a workarounds page.
